src/protocols/multi_party_ecdsa/gg_2020/test.rs

```diff
@@ -527,28 +576,569 @@ fn keygen_t_n_parties(
+            m_b_gamma_vec: m_b_gamma_vec_all[i].clone(),
+            delta_inv: delta_inv.clone(),
+            decom_vec1: decom1_vec.clone(),
+            bc1_vec: bc1_vec.clone(),
+            index: i,
+            sign_keys: sign_keys_vec[i].clone(),
+            s_ttag: ttag,
+        };
+        result_stage5_vec.push(sign_stage5(&input)?);
+    }
+
+    println!("Stage 5 done.");
+    let R_vec: Vec<GE> = result_stage5_vec.iter().map(|a| a.R_vec_i).collect();
+    let R_dash_vec: Vec<GE> = result_stage5_vec.iter().map(|a| a.R_dash_vec_i).collect();
+
+    let mut res_stage6_vec = vec![];
+    for i in 0..ttag {
+        let input = SignStage6Input {
+            R_dash: R_dash_vec[i].clone(),
+            R: R_vec[i].clone(),
+            m_a: m_a_vec[i].0.clone(),
+            e_k: keypair_result.e_vec[s[i]].clone(),
+            k_i: sign_keys_vec[i].k_i.clone(),
+            randomness: m_a_vec[i].1.clone(),
+            party_keys_vec: keypair_result.party_keys_vec.clone(),
+            keypair_result: keypair_result.clone(),
+            index: i,
+            s: s.to_vec(),
+        };
+        let res = sign_stage6(&input)?;
+        res_stage6_vec.push(res);
+    }
+    println!("Stage 6 done.");
+    let mut phase5_proofs_vec = vec![];
+    for i in 0..res_stage6_vec.len() {
+        phase5_proofs_vec.push(res_stage6_vec[i].phase5_proof.clone());
+    }
+    //let phase5_proofs_vec = res_stage6_vec.iter().map(|a| *a.phase5_proof).collect();
+    let mut res_stage7_vec = vec![];
+    for i in 0..ttag {
+        let input = SignStage7Input {
+            phase5_proof_vec: phase5_proofs_vec[i].clone(),
+            R_dash: R_dash_vec[i].clone(),
+            R: R_vec[i].clone(),
+            m_a: m_a_vec[i].0.clone(),
+            ek: keypair_result.e_vec[s[i]].clone(),
+            keypair_result: keypair_result.clone(),
+            s: s.to_vec(),
+            index: i,
+        };
+        let res = sign_stage7(&input)?;
+        res_stage7_vec.push(res);
+    }
+    println!("Stage 7 done.");
+    let message_bn_ = HSha256::create_hash(&[&BigInt::from(bytes_to_sign)]);
+    let mut local_sig_vec = Vec::new();
+    let mut s_vec = Vec::new();
+
+    for i in 0..ttag {
+        let input = SignStage8Input {
+            R_dash_vec: R_dash_vec.clone(),
+            sign_key: sign_keys_vec[i].clone(),
+            message_bn: message_bn_.clone(),
+            R: R_vec[i],
+            sigma: sigma_vec[i],
+            ysum: keypair_result.y_sum.clone(),
+        };
+        let check_local_sig = sign_stage8(&input);
+        if check_local_sig.is_err() {
+            return Err(ErrorType {
+                error_type: format!(" Signature error on index {}", i),
+                bad_actors: vec![],
+            });
+        }
+        let local_sig = check_local_sig.unwrap();
+        s_vec.push(local_sig.s_i.clone());
+        local_sig_vec.push(local_sig);
+    }
+
+    println!("Stage 8 done. s_vec len {}", s_vec.len());
```

Comment:
how do you define a "stage" ? 
the protocol works in 6 phases (not counting aborts), it is understandable at this level of the interface to ignore the actual phases terminology but just curious to what is method of division into steps ? (if you plan on supporting the abort options it should be consistent )